### DISCUSSION TITLE
Add multi task training in ClinicaDL

### DISCUSSION THREAD

**Discussion by ravih18** [aramis-lab/clinicadl#255] [Ideas]
See how to add multi task training in ClinicaDL:
* how to add it to the command line ? (only TOML ?)
* how to write it in the MAPS ?
* how to train several networks / single network ?
* will we use the task manager ? or macro task manager ?
* how to select the correct losses and add it ?

**Comment by 14thibea** [comment]
I think it may be very difficult to add it simply in the commandline. For example if we want to train two classification tasks at the same time, the first one on the diagnosis and the second one on the sex, it means that we will have to link each classification task to a `label`, so it could be easier to use a dictionnary structure / TOML to define it.

**Comment by sophieloiz** [comment]
**Command Line** 
Concerning the way to add it to the command line, I agree with @14thibea , I think that a good start can be to use a TOML 

**MAPS Manager**
In the MAPS Manager, I see two new functions to create: _train_multitask and _ensemble_prediction_multitask

The differences between _train and _train multi task will be to compute N loss (with N the number of tasks) instead of one.  Besides, to evaluate the model when no gradients are accumulated, the new function test_multitask from the task manager will be called. 

**Train single network** 
In the previous version AD-DL, multi-task was implemented training a single network which was divided at one point so that each task has it’s own classifier. This an example of one architecture implemented for two tasks:
`
class Conv5_FC3_MultiTask_2_2(nn.Module):
  
    def __init__(self, dropout=0.5):
        super(Conv5_FC3_MultiTask_2_2, self).__init__()

        self.features = nn.Sequential(
            nn.Conv3d(1, 8, 3, padding=1),
            nn.BatchNorm3d(8),
            nn.ReLU(),
            PadMaxPool3d(2, 2),

            nn.Conv3d(8, 16, 3, padding=1),
            nn.BatchNorm3d(16),
            nn.ReLU(),
            PadMaxPool3d(2, 2),

            nn.Conv3d(16, 32, 3, padding=1),
            nn.BatchNorm3d(32),
            nn.ReLU(),
            PadMaxPool3d(2, 2),

            nn.Conv3d(32, 64, 3, padding=1),
            nn.BatchNorm3d(64),
            nn.ReLU(),
            PadMaxPool3d(2, 2),

            nn.Conv3d(64, 128, 3, padding=1),
            nn.BatchNorm3d(128),
            nn.ReLU(),
            PadMaxPool3d(2, 2),

        )

        self.classifier3 = nn.Sequential(
            Flatten(),
            nn.Dropout(p=dropout),
            nn.Linear(128 * 6 * 7 * 6, 1300),
            nn.ReLU(),
            nn.Linear(1300, 50),
            nn.ReLU(),
            nn.Linear(50, 2)
        )


        self.flattened_shape = [-1, 128, 6, 7, 6]

    def forward(self, x):
        x = self.features(x) # the same for all the tasks
        out1 = self.classifier3(x)
        out2 = self.classifier3(x)

        return out1, out2`
